app/src/main/java/com/bashkevich/androidfundamentals/model/MoviesRepository.kt

```diff
@@ -1,29 +1,54 @@
+import androidx.lifecycle.map
+import com.bashkevich.androidfundamentals.model.entity.MovieActorCrossRef
+import com.bashkevich.androidfundamentals.model.viewobject.Movie
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+
+class MoviesRepository(applicationContext: Context) {
+    val database: MoviesDatabase = MoviesDatabase.create(applicationContext)
+
+    fun getAllMovies() = Transformations.map(database.moviesDao.getAllMovies()) { entities ->
+        entities.map { EntityMapper.fromMovieEntity(it) }
+    }
 
-    suspend fun getTopRatedMovies() = RetrofitModule.moviesApi.getTopRatedMovies()
-    suspend fun getNowPlayingMovies() = RetrofitModule.moviesApi.getNowPlayingMovies()
-    suspend fun getUpcomingMovies() = RetrofitModule.moviesApi.getUpcomingMovies()
-    suspend fun getPopularMovies() = RetrofitModule.moviesApi.getPopularMovies()
-    suspend fun getAllMovies(moviesIds: List<Int>) = moviesIds.distinct().map {
-        EntityMapper.convertToMovieFromDto(RetrofitModule.moviesApi.getMovieById(it))
+    suspend fun findMovieById(movieId: Int) = withContext(Dispatchers.IO) {
+        EntityMapper.fromMovieWithActors(database.moviesDao.getMovieWithActorsById(movieId))
     }
 
-    suspend fun getActors(movieId: Int) =
-        RetrofitModule.moviesApi.getActors(movieId).cast.filter { it.profilePicture != null }
-            .map { EntityMapper.convertActorFromDto(it) }
+    suspend fun refreshMovies() = withContext(Dispatchers.IO) {
+        val nowPlayingMoviesIds =
+            RetrofitModule.moviesApi.getNowPlayingMovies().results.map { it.id }
+        val upcomingMoviesIds = RetrofitModule.moviesApi.getUpcomingMovies().results.map { it.id }
+        val popularMoviesIds = RetrofitModule.moviesApi.getPopularMovies().results.map { it.id }
+        val topRatedMoviesIds = RetrofitModule.moviesApi.getTopRatedMovies().results.map { it.id }
 
-    fun cacheAllMovies(movies: List<Movie>) {
-        allMovies = movies
-    }
+        val allMoviesIds = topRatedMoviesIds.plus(popularMoviesIds).plus(upcomingMoviesIds)
+            .plus(nowPlayingMoviesIds).distinct()
+
+        allMoviesIds.forEach { movieId ->
+
+            val movieEntity =
+                EntityMapper.toMovieEntity(RetrofitModule.moviesApi.getMovieById(movieId))
+
+            database.moviesDao.insertMovie(movieEntity)
+
+            val actorsEntities =
+                RetrofitModule.moviesApi.getActors(movieId).cast.filter { it.profilePicture != null }
+                    .map { EntityMapper.toActorEntity(it) }
 
+            database.moviesDao.insertAllActors(actorsEntities)
 
-    fun findMovieById(movieId: Int) = allMovies?.find { it.id == movieId }
+            val actorsIds = actorsEntities.map { it.actorId }
 
-    companion object {
-        var allMovies: List<Movie>? = null
+            actorsIds.forEach { actorId ->
+                database.moviesDao.insertJoin(MovieActorCrossRef(movieId, actorId))
```

Comment:
how about this? https://stackoverflow.com/questions/44361824/how-can-i-represent-a-many-to-many-relation-with-android-room-when-column-name